replica/table.cc

```diff
@@ -1249,10 +1249,13 @@ void table::set_compaction_strategy(sstables::compaction_strategy_type strategy)
             cg.get_backlog_tracker().copy_ongoing_charges(new_bt, move_read_charges);
 
             new_sstables = make_lw_shared<sstables::sstable_set>(new_cs.make_sstable_set(t._schema));
-            cg.main_sstables()->for_each_sstable([this] (const sstables::shared_sstable& s) {
-                add_sstable_to_backlog_tracker(new_bt, s);
+            std::vector<sstables::shared_sstable> new_sstables_for_backlog_tracker;
+            new_sstables_for_backlog_tracker.reserve(cg.main_sstables()->all()->size());
```

Comment:
Doesn't all build a sstable_list?
If so, we can just keep it and use it here?

Reply:
Not always. It builds a new list if the set is compound (maintenance + main or sets of all groups). Here we're working on main set only of a single group.

For partititioned_sstable_set, for example, we just return its local copy of all.

I think the interface is fragile as user can modify local copy if it doesn't copy on write.

lw_shared_ptr<sstable_list> all() const;

should be changed to

lw_shared_ptr<const sstable_list> all() const;